### PR TITLE
[Issue#52]Support finer-grain control on MySQL threads.

### PR DESCRIPTION
**Pull request by zwang28**
[Issue#52](https://github.com/kunpengcompute/mysql-server/issues/52) Support finer-grain control on MySQL threads.
Added new system variables as follow,
- sched_affinity_foreground_thread
- sched_affinity_log_writer
- sched_affinity_log_flusher
- sched_affinity_log_write_notifier
- sched_affinity_log_flush_notifier
- sched_affinity_log_closer
- sched_affinity_log_checkpointer
- sched_affinity_purge_coordinator

Added new system status Sched_affinity_status.

### CODE REVIEW

File: sql/sched_affinity_manager.cc [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+        numa_bitmask_isbitset(bm_background, i)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool Sched_affinity_manager_numa::check_thread_process_compatibility(
+    bitmask *bm_thread, bitmask *bm_proc) {
+  if (bm_thread == nullptr || bm_proc == nullptr) {
```
Comment (rejohn) [inline sql/sched_affinity_manager.cc:165]: 这个接口和上面的check_foreground_background_compatibility接口实现的功能一致，可考虑合一，减少冗余代码

Comment (zwang28) [reply sql/sched_affinity_manager.cc:165]: for loop里的逻辑略不同

**Review by rejohn — APPROVED** [review]

**Review by rejohn — APPROVED** [review]

File: storage/innobase/log/log0write.cc [hunk 2]
```diff
@@ -2043,6 +2057,13 @@ void log_writer(log_t *log_ptr) {
   }
 
   log_writer_mutex_exit(log);
+
+  if (is_registered_to_sched_affinity && sched_affinity_manager != nullptr) {
```
Comment (liuwei-ck) [inline storage/innobase/log/log0write.cc:2061]: 合并if条件，减少if嵌套

File: sql/sched_affinity_manager.cc [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+}
+
+bool Sched_affinity_manager_numa::bind_to_group(const pid_t pid) {
+  auto thread_type = get_thread_type_by_pid(pid);
+  if (thread_type == Thread_type::UNDEFINED) {
+    return false;
+  }
+  if (!is_thread_sched_enabled(thread_type)) {
+    return false;
+  }
```
Comment (liuwei-ck) [inline sql/sched_affinity_manager.cc:224]: 都返回false ，if语句可以合并

File: sql/sched_affinity_manager.cc [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+  auto ret = numa_sched_setaffinity(
+      pid, sched_affinity_group[best_index].avail_cpu_mask);
+  if (ret == 0) {
+    ++sched_affinity_group[best_index].assigned_thread_num;
+    m_pid_group_id[pid] = best_index;
+    return true;
+  } else {
+    return false;
```
Comment (liuwei-ck) [inline sql/sched_affinity_manager.cc:252]: 这个 else无意义。可以不需要。直接返回false

**Review by liuwei-ck — APPROVED** [review]

**Pull request merged**